File: python/03_functions.py
"""
Lesson 03 — Functions

Python: def, parameters, defaults, *args, **kwargs, returns
Rust:   fn, typed parameters, -> return type, no defaults/varargs in this lesson
"""


# --- 1. Basic function: no args, no return ---
def greet():
    print("Hello!")


greet()


# --- 2. Function with parameters ---
def greet_person(name, greeting="Hello"):
    """A simple function with a default argument."""
    print(f"{greeting}, {name}!")


greet_person("Siseng")
greet_person("Siseng", "Good morning")
greet_person("Ricky", greeting="Hi")     # keyword args


# --- 3. Function with a return value ---
def add(a, b):
    return a + b


result = add(3, 4)
print(f"3 + 4 = {result}")


# --- 4. Multiple return values (Python: tuple) ---
def min_and_max(numbers):
    return min(numbers), max(numbers)


lo, hi = min_and_max([3, 1, 4, 1, 5, 9, 2, 6])
print(f"min={lo}, max={hi}")


# --- 5. *args and **kwargs ---
def tag(tag_name, *content, **attrs):
    """Render an HTML-ish tag. *content = positional, **attrs = keyword."""
    attr_str = " ".join(f'{k}="{v}"' for k, v in attrs.items())
    inner = " ".join(content)
    print(f"<{tag_name} {attr_str}>{inner}</{tag_name}>")


tag("a", "click me", href="https://example.com", target="_blank")


# --- 6. Type hints (Python: optional, not enforced) ---
def square(num: int) -> int:
    """Square a number. Type hints are documentation only at runtime."""
    return num * num


print(f"square(5) = {square(5)}")

# This CALLS in Python even though the type hint is `int`.
# Python's type hints are NOT enforced at runtime — the error only fires inside:
try:
    square("hi")
except TypeError as e:
    print(f"Python: caught TypeError → '{e}'")
print("  (Rust would have refused this at compile time, no runtime error needed)")


# --- 7. Functions are first-class objects ---
def apply(f, x):
    """Pass a function as an argument."""
    return f(x)


print(f"apply(square, 7) = {apply(square, 7)}")

File: rust/03_functions.rs
// Lesson 03 — Functions
// Standalone .rs (compile with: rustc rust/03_functions.rs -o /tmp/h && /tmp/h)
//
// Key Rust rules in this lesson:
//   - fn name(params) -> return_type { body }
//   - Parameter types REQUIRED (no inference for params)
//   - No default arguments in plain fn (use trait/impl tricks later)
//   - No *args/**kwargs (use slices or macros)
//   - Last expression (no semicolon) is the return value
//   - Multiple returns: use a tuple

fn main() {
    // 1. Basic function call
    greet();

    // 2. Function with parameters
    greet_person("Siseng", "Hello");

    // 3. Function with a return value
    let result = add(3, 4);
    println!("3 + 4 = {result}");

    // 4. Multiple return values: tuple
    let (lo, hi) = min_and_max(&[3, 1, 4, 1, 5, 9, 2, 6]);
    println!("min={lo}, max={hi}");

    // 5. No *args/**kwargs — use slices instead
    print_all(&["a", "b", "c"]);

    // 6. Functions as values (function pointers)
    let answer = apply(square, 7);
    println!("apply(square, 7) = {answer}");
}

// 1. Basic function: no args, no return (unit type `()`)
fn greet() {
    println!("Hello!");
}

// 2. Function with parameters — types REQUIRED
fn greet_person(name: &str, greeting: &str) {
    println!("{greeting}, {name}!");
}

// 3. Function with a return value — last expression (no `;`) is the return
fn add(a: i32, b: i32) -> i32 {
    a + b         // ← no semicolon! this is the return value
    // If you put `;` here, it becomes a statement and the function returns ()
}

// 4. Multiple return values: return a tuple
fn min_and_max(numbers: &[i32]) -> (i32, i32) {
    let mut lo = numbers[0];
    let mut hi = numbers[0];
    for &n in numbers {
        if n < lo { lo = n; }
        if n > hi { hi = n; }
    }
    (lo, hi)       // tuple — returned as the last expression
}

// 5. Variable number of args — Rust uses slices, not varargs
fn print_all(items: &[&str]) {
    for item in items {
        print!("{item} ");
    }
    println!();
}

// 6. Type-annotated function (mandatory in Rust, unlike Python)
fn square(num: i32) -> i32 {
    num * num
}

// 7. Functions are values — pass them as arguments
//    `fn(i32) -> i32` is the type of a function pointer
fn apply(f: fn(i32) -> i32, x: i32) -> i32 {
    f(x)
}
